### ISSUE TITLE
How to deploy a self-developed plugin in webSpoon?

### ISSUE THREAD

**Issue by nazeri84** [HiromuHota/pentaho-kettle#84]
When using the spoon.war file deployed as a servlet in Tomcat on a Linux server, how do you deploy a plugin? In Desktop Spoon you put it into a subfolder of the plugins folder, but where exactly do you place it in the webSpoon folder structure? I don't get it to work following your suggestions here: https://github.com/HiromuHota/pentaho-kettle#plugins

Thank you so much in advance,
Ali

**Comment by HiromuHota** [comment]
Sorry to confuse you, the following directory structure is the most straight-forward.
Under the `CATALINA_HOME` folder, you should already have `plugins` folder next to `webapps`.
Place your plugin folder in that `plugins` folder.

```
CATALINA_HOME
├── bin
│   ├── bootstrap.jar
│   ├── ...
├── conf
│   ├── Catalina
│   ├── catalina.policy
│   ├── ...
├── plugins
│   ├── VerticaBulkLoader
│   ├── elasticsearch-bulk-insert-plugin
│   ├── YourPlugin
├── webapps
│   ├── ROOT
│   ├── docs
│   ├── examples
│   ├── host-manager
│   ├── manager
│   ├── spoon
│   └── spoon.war
```

**Comment by nazeri84** [comment]
As you said I put my plugins in this directory : /opt/tomcat/plugins/
and also I got a test with this /opt/tomcat/plugins/steps/ as well, but it does not work.
based on issue #14 and also #26 I set the environment variable in kettle.properties file by following form:
KETTLE_PLUGIN_BASE_FOLDERS=/opt/tomcat/plugins
but my problem has not be solved yet!
Do you have any idea about this situation?

Thanks again

**Comment by nazeri84** [comment]
I found out my problem is related to CATALINA_HOME variable because it was not set.
I used these commands in the .bashrc file:
export JAVA_HOME=/usr/lib/path/to/java
export CATALINA_HOME=/path/to/tomcat
and the problem has been fixed.

**Comment by HiromuHota** [comment]
Sorry I was on a longer weekend, but happy to hear that.

For the future record, CATALINA_HOME is already set in a Docker container.

**Comment by HiromuHota** [comment]
Actually, your symptom looks a bit strange to me.
Please tell me the following information for other people who would have the same issue in the future:
- What is the Servlet container? Is it a vanilla Tomcat, or the Tomcat in the Pentaho Server?
- Assuming it is Tomcat, how do you start Tomcat (e.g., `./bin/startup.sh`) and what is the current directory when you start it (e.g., `/path/to/tomcat`)?

**Comment by nazeri84** [comment]
I have installed tomcat directly, not that one which exists in Pentaho Server.
to start tomcat service I use this command : 
systemctl start tomcat.service

**Comment by HiromuHota** [comment]
I tested on CentOS7 and plugins in `/usr/share/tomcat/plugins` worked without any issue.
It is probably because I installed Tomcat via Yum (7.0.76 by default) and you seemed to install Tomcat manually.
Tomcat installed via Yum already has CATALINA_HOME, so no problem.

For the record, I needed to change owner of `/usr/share/tomcat` to tomcat:tomcat to avoid permission errors in creating `.kettle/kettle.properties` and `.pentaho/metastore`.
